File: kata-solutions/vending-machine/vending-machine-typescript/src/vending_machine.ts
export enum Choice {
  Cola = "Cola choice",
  FizzyOrange = "Fizzy orange choice",
  Beer = "BEER choice"
}

export enum Can {
  Nothinbg = "No can",
  Coke = "Can of Coke",
  Fanta = "Can of Fanta"
}

class Cashier {
  constructor(private balanceInCents: number = 0) {}  

  public insert(amountInCents: number) {
    this.balanceInCents += amountInCents
  }

  public doesBalanceAllow(priceInCents: number) {
    return this.balanceInCents >= priceInCents
  }

  public buy(amountInCents: number) {
    this.balanceInCents -= amountInCents
  }
}

class Drawer {
  constructor(public can: Can, public priceInCents: number) {}

  deliver(cashier: Cashier) {
    if (!cashier.doesBalanceAllow(this.priceInCents))
      return Can.Nothinbg
    
    cashier.buy(this.priceInCents)
    return this.can    
  }
}

export class VendingMachine {
    private choiceDrawerMap: Map<Choice, Drawer> = new Map<Choice, Drawer>()
    private cashier: Cashier = new Cashier()

    public configure(choice: Choice, can: Can, priceInCents = 0): void {
      this.choiceDrawerMap.set(choice, new Drawer(can, priceInCents))
    }

    public insert(amountInCents: number): void  {
      this.cashier.insert(amountInCents)
    }
  
    public deliver(choice: Choice): Can {
      if (!this.choiceDrawerMap.has(choice))
          return Can.Nothinbg

      let drawer = this.choiceDrawerMap.get(choice) as Drawer
      return drawer.deliver(this.cashier)
    }
}

File: kata-solutions/vending-machine/vending-machine-java/app/src/main/java/VendingMachine.java
import java.util.Map;
import java.util.HashMap;

public class VendingMachine {
    // private Map<Choice, Can> choiceCanMap = new HashMap<Choice, Can>();
    // private Map<Choice, Integer> choicePriceMap = new HashMap<Choice, Integer>();
    // private int balanceInCents = 0;
  
    // public void configure(final Choice choice, final Can can, final int priceInCents) {
    //   this.choiceCanMap.put(choice, can);
    //   this.choicePriceMap.put(choice, priceInCents);
    // }
  
    // public void configure(final Choice choice, final Can can) {
    //   configure(choice, can, 0);
    // }

    // public void insert(final int amountInCents) {
    //   this.balanceInCents = amountInCents;
    // }

    // public Can deliver(final Choice choice) {
    //   if (!choiceCanMap.containsKey(choice)) return Can.NOTHING;

    //   final int canPrice = choicePriceMap.get(choice);
    //   if (canPrice > balanceInCents ) return Can.NOTHING;
      
    //   balanceInCents -= canPrice;
    //   return choiceCanMap.get(choice);
    // }

  private Map<Choice, Drawer> choiceDrawerMap = new HashMap<Choice, Drawer>();
  private final Cashier cashier = new Cashier();

  public void configure(Choice choice, Drawer drawer) {
    this.choiceDrawerMap.put(choice, drawer);
  }

  public void insert(int priceInCents) {
    cashier.insert(priceInCents);
  }
  
  public Can deliver(final Choice choice) {
    if (!this.choiceDrawerMap.containsKey(choice)) return Can.NOTHING;

    return this.choiceDrawerMap.get(choice).deliver(cashier);
  }
  
  public class Cashier {
    private int balanceInCents = 0;
  
    public void insert(final int amountInCents) {
      balanceInCents += amountInCents;
    }
  
    public boolean doesBalanceAllow(final int priceInCents) {
      return balanceInCents >= priceInCents;
    }
  
    public void buy(final int amountInCents) {
      balanceInCents -= amountInCents;
    }
  }
  
  public class Drawer {
    public final Can can;
    public final int priceInCents;

    public Drawer(Can can) {
      this(can, 0);
    }

    public Drawer(Can can, int priceInCents) {
      this.can = can;
      this.priceInCents = priceInCents;
    }

    public Can deliver(final Cashier cashier) {
      if (!cashier.doesBalanceAllow(priceInCents))
        return Can.NOTHING;
    
      cashier.buy(priceInCents);
      return can;
    }
  }
}
